attachment_s3/models/ir_attachment.py

```diff
@@ -0,0 +1,164 @@
+                    'If you want to write in the %s S3 bucket, this variable '
+                    'must be set as well:\n'
+                    '* AWS_BUCKETNAME\n'
+                    'Optionally, the S3 host can be changed with:\n'
+                    '* AWS_HOST\n'
+                    ) % (bucket_name, bucket_name)
+
+            raise except_orm(_('Configuration Error'), msg)
+
+        try:
+            conn = connect_s3(aws_access_key_id=access_key,
+                              aws_secret_access_key=secret_key)
+
+        except S3ResponseError as error:
+            # log verbose error from s3, return short message for user
+            _logger.exception('Error during connection on S3')
+            raise except_orm(_('S3 Error'), self._parse_s3_error(error))
+
+        bucket = conn.lookup(bucket_name)
+        if not bucket:
+            bucket = conn.create_bucket(bucket_name)
+        return bucket
+
+    @staticmethod
+    def _parse_s3_error(s3error):
+        msg = s3error.reason
+        # S3 error message is a XML message...
+        doc = xml.dom.minidom.parseString(s3error.body)
+        msg_node = doc.getElementsByTagName('Message')
+        if msg_node:
+            msg = '%s: %s' % (msg, msg_node[0].childNodes[0].data)
+        return msg
+
+    def _store_file_read(self, fname, bin_size=False):
+        if fname.startswith('s3://'):
+            s3uri = S3Uri(fname)
+            try:
+                bucket = self._get_s3_bucket(name=s3uri.bucket())
+            except except_orm:
+                _logger.exception(
+                    "error reading attachment '%s' from object storage", fname
+                )
+                return ''
+            filekey = bucket.get_key(s3uri.item())
+            if filekey:
+                read = base64.b64encode(filekey.get_contents_as_string())
+            else:
+                read = ''
+                _logger.info(
+                    "attachment '%s' missing on object storage", fname
+                )
+            return read
+        else:
+            return super(IrAttachment, self)._store_file_read(fname, bin_size)
+
+    def _store_file_write(self, storage, key, bin_data):
+        if storage == 's3':
+            bucket = self._get_s3_bucket()
+            filekey = bucket.get_key(key) or bucket.new_key(key)
+            filename = 's3://%s/%s' % (bucket.name, key)
+            try:
+                filekey.set_contents_from_string(bin_data)
+            except S3ResponseError as error:
+                # log verbose error from s3, return short message for user
+                _logger.exception(
+                    'Error during storage of the file %s' % filename
+                )
+                raise except_orm(
+                    _('S3 Error'),
+                    _('The file could not be stored: %s') %
+                    (self._parse_s3_error(error),)
+                )
+        else:
+            _super = super(IrAttachment, self)
```

Comment:
why add this in variable but not directly use beside ?

Reply:
Most of the code is actually a copy of the v9 and is working out of the box. As I had a tight schedule, I didn't take the time to refactor nor optimize the existing code.

Reply:
@Tonow-c2c Line would be too long